### ISSUE TITLE
Make librad executor agnostic

### ISSUE THREAD

**Issue by kim** [radicle-dev/radicle-link#61]
## Problem

There are a few cases where it would be good to have control over the executor / event loop when dealing with async code. For example, blocking code which needs to call into async code (but obviously needs to block on the result) should be passed some object onto which to spawn tasks, instead of relying on a magically available (thread-local) executor when it so happens that the outermost `main` starts one. Another example are background / periodic tasks.

The issue here is the ecosystem fragmentation between `futures`, `tokio` and `async-std`, which, honestly, I don't perceive as friendly (especially from the `tokio` side). In other words, there is no single interface for a thing that can be passed around, and onto which the user can spawn a `Future`. This implies that, as a library author, I am forced to make a choice, where it _seems_ like the right thing to do would be to leave the choice between `tokio` and `async-std` to the application author. The fact that `futures` also comes with executors only makes it more confusing.

## Solution

I don't have a solution.

**Comment by kim** [comment]
Perhaps @geigerzaehler can tell us about ways to deal with this?

**Comment by kim** [comment]
We should also include a `rant` label. But seriously, it is incomprehensible to me why `tokio` 0.2 introduces the exact same trait signatures as `futures`, but is not just re-exporting `futures`.

**Comment by geigerzaehler** [comment]
I don’t have a solution to your particular problem (passing a task executor) either. I’m not aware of a common task executor trait.

In the registry client we have a related issue: We’re dealing with futures provided by a library (jsonrpc) and provide a futures API ourselves. The user of our library needs to provide an appropritate runtime (tokio 0.1) to use our APIs. Since this may not always be desirable we also allow the user to create a client that has its own executor.

https://github.com/radicle-dev/radicle-registry/blob/9e7fe9a9118775eae37d7ccef551c90819790d43/client/src/backend/remote_node_with_executor.rs#L36-L39

With this approach the are no requirements on an existing runtime. However, if the user does run another runtime (which they need because we’re dealing with futures after all) then the two runtimes might impact performance.

I don’t know how relevant this is to your problem but it might give you some insights.

**Comment by kim** [comment]
This may be what I'm looking for: 

https://blog.wnut.pw/2020/02/25/anouncing-async_executors-a-building-block-for-executor-agnostic-libraries/

(via @massimiliano-mantione)

**Comment by kim** [comment]
Oh well, I'm not sure. There's two things I need:

1. Spawn a "background task", which should be cancelled when the struct representing a stateful computation gets dropped

    afaik, this can't directly be done with neither `futures`, `tokio`, nor `async-std`. What I did was to `tokio::spawn` them and have them check on every iteration of a `loop` if the "parent" type was dropped. Apparently, `async_executors` [can do that](https://docs.rs/async_executors/0.2.0/async_executors/struct.JoinHandle.html)

2. Call into blocking code from async code, which in turn will invoke async code again, pretending it is blocking (it is using the FFI)

    The solution I came up with is to use [`tokio::task::block_in_place`](https://docs.rs/tokio/0.2.13/tokio/task/fn.block_in_place.html), which will make it so the blocking code runs on its own thread, not interfering with any other futures (obviously, this requires a threaded runtime). The blocking code in turn uses [`futures::executor::block_on`](https://docs.rs/futures/0.3.4/futures/executor/fn.block_on.html), which will run a lightweight single-threaded executor on the current thread to schedule just that task. This seems to be what I want conceptually, but requires to make a choice between `tokio` and `async-std` -- `async_executors` doesn't provide an API for "blocking spawn".

**Comment by kim** [comment]
I played around with `async_executors`, and it's not really practical at this point: the `Spawn` handle has to be passed around explicitly (which I very much agree with), but library users will typically want to use `#[tokio::main]` or somesuch, which creates a global runtime. Now, it's possible to get at a `Handle` of the implicit runtime, but `async_executors` doesn't provide a way to convert this `Handle` without also having access to the runtime itself.

**tl;dr** We are currently stuck with tokio, but in a silent way (it's not visible in the types). This sucks.
